paddle/fluid/framework/details/graph_print_pass.cc

```diff
@@ -0,0 +1,150 @@
+  void AddCustomEdge(const Callback& cb) {
+    stream_ << cb() << std::endl;
+  }
+
+ private:
+  std::ostringstream stream_;
+};
+
+template <typename T, typename Container>
+std::vector<T*> FilterByNodeWrapper(const Container& con) {
+  std::vector<T*> ret;
+  for (auto& node : con) {
+    auto i = dynamic_cast<T*>(node.get());
+    if (i != nullptr) ret.emplace_back(i);
+  }
+  return ret;
+}
+
+std::unordered_map<ir::Node*, int> SSAGraphPrinterImpl::ToGraphvizNode(
+    const ir::Graph& graph) const {
+  // Convert to GraphvizNode format
+  auto& graphviz_nodes = graph.Get<GraphvizNodes>(kGraphviz);
+  graphviz_nodes.clear();
+  std::unordered_map<ir::Node*, int> vars;
+  std::unordered_map<ir::Node*, GraphvizOp*> ops;
+  int var_id = 0;
+  int op_id = 0;
+  for (auto& node : graph.Nodes()) {
+    if (node->IsVar()) {
+      graphviz_nodes.emplace(new GraphvizVar(node, var_id));
+      vars.emplace(std::make_pair(node, var_id++));
+    } else if (node->IsOp()) {
+      std::unique_ptr<GraphvizOp> op(new GraphvizOp(node, op_id++));
+      ops[node] = op.get();
+      graphviz_nodes.emplace(std::move(op));
+    } else {
+      PADDLE_THROW("Unknown op type");
+    }
+  }
+
+  // Detect circle. Draw circle in different lines
+  std::vector<std::vector<ir::Node*>> circles;
+  const std::string kCircleEdge = "[color=red,penwidth=3.0]";
+  if (ir::FindCircleSubGraph(graph, &circles)) {
+    VLOG(3) << "Graph has circle! circles count : " << circles.size();
+    for (auto& circle : circles) {
+      for (size_t i = 0; i < circle.size() - 1; ++i) {
+        GraphvizOp* prev = ops[circle[i]];
+        GraphvizOp* next = ops[circle[i + 1]];
+        std::string prev_op = "op_" + std::to_string(prev->Id());
+        std::string next_op = "op_" + std::to_string(next->Id());
+        prev->AddCustomEdge([&]() -> std::string {
+          return prev_op + "->" + next_op + kCircleEdge;
+        });
+      }
+    }
+  }
+  return vars;
+}
+
+void SSAGraphPrinterImpl::Print(const ir::Graph& graph,
+                                std::ostream& sout) const {
+  auto vars = ToGraphvizNode(graph);
+  auto& nodes = graph.Get<GraphvizNodes>(kGraphviz);
+
+  sout << "digraph G {\n";
+  for (auto& var : FilterByNodeWrapper<GraphvizVar>(nodes)) {
+    sout << *var;
+  }
+
+  for (auto& op : FilterByNodeWrapper<GraphvizOp>(nodes)) {
+    op->AddEdge([&vars](ir::Node* var) { return vars.at(var); });
+    sout << *op;
+  }
+  sout << "}\n";
+}
+
+std::unique_ptr<ir::Graph> SSAGraphPrintPass::ApplyImpl(
+    std::unique_ptr<ir::Graph> graph) const {
+  printer_.reset(new SSAGraphPrinterImpl());
+  std::unique_ptr<std::ostream> fout(
+      new std::ofstream(Get<std::string>(kGraphvizPath)));
+  PADDLE_ENFORCE(fout->good() == true, "Failed to open file.");
+
+  printer_->Print(*graph, *fout);
+  return graph;
+}
+
+}  // namespace details
+}  // namespace framework
+}  // namespace paddle
+
+REGISTER_PASS(graph_print_pass, paddle::framework::details::SSAGraphPrintPass)
```

Comment:
Add a config in graph_viz_pass to support this type of sort, if the core difference is some details not the graphviz itself?

The graphviz related implementation can still be reused, there are some color and theme customization functions in graph_viz_pass, and that works fine, no need to re-implement one.

Reply:
The inference relays heavily on `graph_viz_pass` to visualize the graph, and if the inplace-pass switch on in the inference, the `graph_viz_pass` can't works.

Better to modify the existing `graph_viz_pass` so that the inference can reuse it.